velox/exec/tests/VectorHasherTest.cpp

```diff
@@ -693,3 +694,89 @@ TEST_F(VectorHasherTest, computeValueIdsStrings) {
 
   ASSERT_LE(uniqueValues.size(), multiplier);
 }
+
+namespace {
+template <typename T>
+// Appends a null for id 0, min of T for 1, max of T for 2 and 0 for 3.
+void append(FlatVectorPtr<T>& vector, int32_t id) {
+  auto size = vector->size();
+  vector->resize(size + 1);
+  if (id == 0) {
+    vector->setNull(size, true);
+  } else {
+    vector->set(
+        size,
+        id == 1       ? std::numeric_limits<T>::min()
+            : id == 2 ? std::numeric_limits<T>::max()
+                      : 0);
+  }
+}
+} // namespace
+
+TEST_F(VectorHasherTest, endOfRange) {
+  // We sample with large but not end-reaching ranges of different ints.
+  // We check that ranges are rounded up to ends of the data types.
+  // We check that nulls, min and max values in each column combine to make
+  // distinct normalized keys.
+  constexpr int64_t kTinyRange = std::numeric_limits<uint8_t>::max();
+  constexpr int64_t kSmallRange = std::numeric_limits<uint16_t>::max();
+  constexpr int64_t kIntRange = std::numeric_limits<uint32_t>::max();
+
+  using exec::VectorHasher;
+  auto tinySample = vectorMaker_->flatVector(std::vector<int8_t>{-100, 110});
+  auto smallSample =
+      vectorMaker_->flatVector(std::vector<int16_t>{-30000, 31000});
+  auto intSample =
+      vectorMaker_->flatVector(std::vector<int32_t>{-2001000000, 2000000000});
+  auto tinyHasher = VectorHasher::create(TINYINT(), 0);
+  auto smallHasher = VectorHasher::create(SMALLINT(), 1);
+  auto intHasher = VectorHasher::create(INTEGER(), 2);
+  raw_vector<uint64_t> result;
+  SelectivityVector rows(2);
+  tinyHasher->computeValueIds(*tinySample, rows, result);
+  smallHasher->computeValueIds(*smallSample, rows, result);
+  intHasher->computeValueIds(*intSample, rows, result);
+  uint64_t distinct;
+  uint64_t range;
+  tinyHasher->cardinality(50, range, distinct);
+  // Distinct count is 4 because 2 values + 50% is 3 and 1 for the null value.
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kTinyRange + 2);
+  smallHasher->cardinality(50, range, distinct);
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kSmallRange + 2);
+  intHasher->cardinality(50, range, distinct);
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kIntRange + 2);
+  auto multiplier1 = tinyHasher->enableValueRange(1, 50);
+  auto multiplier2 = smallHasher->enableValueRange(multiplier1, 50);
+  auto multiplier3 = intHasher->enableValueRange(multiplier2, 50);
+  // The total combinations are the product of the ranges + 1 for null.
+  EXPECT_EQ(
+      (kTinyRange + 2) * (kSmallRange + 2) * (kIntRange + 2), multiplier3);
```

Comment:
nit: This line of code and the comment above might be a bit confusing because the comment says "...the product of the ranges + 1..." while the code does `kTinyRange + 2`. Maybe add 1 to the definition of kTinyRange, kSmallRange, and kIntRange, and then plus 1 at other places?